feat: add new move count metric (#1072)
TimefoldAI/timefold-solver: docs/src/modules/ROOT/pages/constraints-and-score/performance.adoc

[#scoreCalculationPerformanceTricksOverview]
== Overview

The `Solver` will normally spend most of its execution time evaluating moves and running score calculation,
which is called in its deepest loops.

> Faster move evaluation will return the same solution in less time with the same algorithm,

Reviewer: In this case, we are arguably talking about faster score calculation.
That is the deciding factor for speed, if everything else is equal. Move evaluation speed is _determined_ by score calculation speed.

which normally means a better solution in equal time.

> [#moveEvaluationSpeed]

Reviewer: Have you checked all the references to this?
Antora has a habit of failing when refs don't exist.

Also, I'd call this section "Move evaluation and score calculation speed".

Author: Yes, I did check the references.

== Move evaluation speed

After solving a problem, the `Solver` will log the __move evaluation speed per second__.
This is a good measurement of Move evaluation performance,
despite that it is affected by non-score calculation execution time.
It depends on the problem scale of the problem dataset.
Normally, even for large scale problems, it is higher than ``1000``,
except if you are using xref:constraints-and-score/score-calculation.adoc#easyScoreCalculation[``EasyScoreCalculator``].

[IMPORTANT]
====

> When improving your move evaluation, focus on maximizing the move evaluation speed,

Reviewer: See, I don't agree with this statement.

To improve your performance, you have to increase your score calculation speed - if you focus on move evaluation speed, it means you start removing slow moves, and that is counter-productive. You need to increase the score calculation speed while _keeping_ the existing moves.

Comparing the best score with the original's best score is pointless: it's comparing apples and oranges.
====

The solver usually calculates the score for each move it evaluates.
This means a direct relationship exists between the score calculation and the evaluated move, typically a `1:1` ratio.
However, some moves,
such as xref:optimization-algorithms/move-selector-reference.adoc#ruinRecreateMoveSelector[Ruin and Recreate],
involve processing the score multiple times within the same move.

The use of moves like Ruin and Recreate may increase the number of score calculations,

> which makes the relationship `1:1` invalid.

Reviewer: ```suggestion
making this ratio `N:1`.
```

In such cases,
the metric xref:using-timefold-solver/benchmarking-and-tweaking.adoc#benchmarkReportMoveEvaluationSpeedSummary[__move evaluation speed per second__] makes more sense,

> as it will count the number of moves individually.

Reviewer: ```suggestion
as the move count will stay stable no matter how many score calculations the moves perform.
```

== `stepLimit` benchmark

Not all score constraints have the same performance cost.

> Sometimes one score constraint can kill the move evaluation performance outright.

Reviewer: ```suggestion
Sometimes one score constraint can performance outright.
```
